x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearnToRankService.java

```diff
@@ -0,0 +1,195 @@
+        ScriptService scriptService,
+        XContentParserConfiguration parserConfiguration
+    ) {
+        this.modelLoadingService = modelLoadingService;
+        this.scriptService = scriptService;
+        this.trainedModelProvider = trainedModelProvider;
+        this.parserConfiguration = parserConfiguration;
+    }
+
+    /**
+     * Asynchronously load a regression model to be used for learn to rank.
+     *
+     * @param modelId The model id to be loaded.
+     * @param listener Response listener.
+     */
+    public void loadLocalModel(String modelId, ActionListener<LocalModel> listener) {
+        modelLoadingService.getModelForLearnToRank(modelId, listener);
+    }
+
+    /**
+     * Asynchronously load the learn to rank config by model id.
+     * Once the model is loaded, templates are executed using params provided.
+     *
+     * @param modelId Id of the model.
+     * @param params Templates params.
+     * @param listener Response listener.
+     */
+    public void loadLearnToRankConfig(String modelId, Map<String, Object> params, ActionListener<LearnToRankConfig> listener) {
+        trainedModelProvider.getTrainedModel(
+            modelId,
+            GetTrainedModelsAction.Includes.all(),
+            null,
+            ActionListener.wrap(trainedModelConfig -> {
+                if (trainedModelConfig.getInferenceConfig() instanceof LearnToRankConfig retrievedInferenceConfig) {
+                    for (LearnToRankFeatureExtractorBuilder builder : retrievedInferenceConfig.getFeatureExtractorBuilders()) {
+                        builder.validate();
+                    }
+                    listener.onResponse(applyParams(retrievedInferenceConfig, params));
+                    return;
+                }
+                listener.onFailure(
+                    ExceptionsHelper.badRequestException(
+                        Messages.getMessage(
+                            Messages.INFERENCE_CONFIG_INCORRECT_TYPE,
+                            Optional.ofNullable(trainedModelConfig.getInferenceConfig()).map(InferenceConfig::getName).orElse("null"),
+                            LearnToRankConfig.NAME.getPreferredName()
+                        )
+                    )
+                );
+            }, listener::onFailure)
+        );
+    }
+
+    /**
+     * Applies templates params to a {@link LearnToRankConfig} object.
+     *
+     * @param config Original config.
+     * @param params Templates params.
+     * @return A {@link LearnToRankConfig} object with templates applied.
+     *
+     * @throws IOException
+     */
+    private LearnToRankConfig applyParams(LearnToRankConfig config, Map<String, Object> params) throws IOException {
+        if (scriptService.isLangSupported(DEFAULT_TEMPLATE_LANG) == false) {
+            return config;
+        }
+
+        if (params == null || params.isEmpty()) {
+            // TODO: better handling of missing parameters.
```

Comment:
ℹ️ Currently when a param is missing it will be kept intact in the final query feature extractor.

Because the features map will be ultimately casted to an array of `double`, I think the best options is to extract a feature equals to `0` in this case. To achieve this we can return a query extractor using a `match_none` query if a params is missing.

I would prefer opening a follow-up PR to address this specific issue (because it is non-trivial to add the detection of missing parameters to the MustacheScriptEngine).

Reply:
++ to a new PR. It would be nice to expose a "default" option for the user to set. There will be cases where 0.0 is not an appropriate default for missing values.